timsort.hpp

```diff
@@ -126,34 +204,299 @@ template <typename Value, typename LessFunction> class Compare {
+        assert(len > 0);
+        value_t *nend = base_.startp_ + len;
+        if (nend > base_.alloc_limitp_) {
+            // Our current allocation is too small so allocate a
+            // new array entirely
+            LengthType const new_cap = std::max(base_.next_capacity(), len);
+            destruct();
+            value_t *nv;
+            try {
+                nv = base_.allocate(new_cap);
+            } catch (...) {
+                base_.startp_ = 0;
+                base_.endp_ = 0;
+                base_.alloc_limitp_ = 0;
+                throw;
+            }
+            base_.startp_ = nv;
+            base_.alloc_limitp_ = nv + new_cap;
+            nend = nv + len;
+            assert(nend <= base_.alloc_limitp_);
+        }
+        // The most correct way to copy these trivially-constructable elements
+        // into base_.startp_ would be to do:
+        //   GFX_TIMSORT_MOVE_RANGE(in_begin, in_begin + len, base_.startp_);
+        // Since the types are default-constructable the STL will be able to
+        // reduce this to a call to std::memmove()   However, we know that we're
+        // never dealing with overlapping memory here, so it's a tiny bit
+        // faster to use std::memcpy() instead.
+        std::memcpy(base_.startp_, &*in_begin, len * sizeof(base_.startp_[0]));
+        base_.endp_ = nend;
+    }
+};
+#endif
+
+// Dynamically-allocated stack of pending "runs" that TimSort needs to merge.
+template <typename RandomAccessIterator, typename LengthType> class TimSortRunStack {
+  public:
+    struct run {
+        RandomAccessIterator base;
+        LengthType len;
+    };
+  private:
+    struct contents : public std::allocator<struct run> {
+        contents() : startp_(0), endp_(0), alloc_limitp_(0) {
+        }
+        struct run *startp_;
+        struct run *endp_;
+        struct run const *alloc_limitp_;
+
+        struct run *push_back_grow() {
+            assert(endp_ == alloc_limitp_);
+            LengthType const old_cap = (endp_ - startp_);
+            LengthType const new_cap = (startp_ == 0) ? 16 : ((old_cap * 3) >> 1);
+            assert(new_cap > old_cap);
+            struct run *nv = this->allocate(new_cap);
+            if (startp_ != 0) {
+                std::copy(startp_, endp_, nv);
+                this->deallocate(startp_, endp_ - startp_);
+            }
+            startp_ = nv;
+            // endp_ is set by our caller
+            alloc_limitp_ = nv + new_cap;
+            return nv + old_cap + 1;
+        }
+    };
+    struct contents c_;
+#if GFX_TIMSORT_USE_CXX11
+    TimSortRunStack(const TimSortRunStack&) = delete;
+    TimSortRunStack& operator=(const TimSortRunStack&) = delete;
+#endif
+  public:
+    TimSortRunStack() {
+    }
+    ~TimSortRunStack() {
+        c_.deallocate(c_.startp_, c_.alloc_limitp_ - c_.startp_);
+    }
+    void push_back(RandomAccessIterator const runBase, LengthType const runLen) {
+        struct run *nend = c_.endp_ + 1;
```

Comment:
FYI, this results in undefined behavior the first time around, since `c_.endp == nullptr` due to being initialized by `contents() : startp_(0), endp_(0), alloc_limitp_(0)`

I think the easiest fix is just to check `if (c_.endp_ == c_.alloc_limitp_)` and compute `c_.endp_ + 1` only on the `else` branch.